File: backend/models/library.py
from datetime import datetime
from extensions import db

# Association table for Many-to-Many relationship between Books and Authors
book_authors = db.Table('book_authors',
    db.Column('book_isbn', db.String(13), db.ForeignKey('books.isbn', ondelete='CASCADE'), primary_key=True),
    db.Column('author_id', db.Integer, db.ForeignKey('authors.id', ondelete='CASCADE'), primary_key=True)
)

class Publisher(db.Model):
    __tablename__ = 'publishers'
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(255), nullable=False)
    address = db.Column(db.Text, nullable=False)
    contact_email = db.Column(db.String(255), nullable=False)

    books = db.relationship('Book', backref='publisher_ref', cascade='all, delete-orphan')
    orders = db.relationship('PurchaseOrder', backref='publisher_ref', cascade='all, delete-orphan')

class Author(db.Model):
    __tablename__ = 'authors'
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(255), nullable=False)
    biography = db.Column(db.Text, nullable=True)
    date_of_birth = db.Column(db.Date, nullable=True)

class Category(db.Model):
    __tablename__ = 'categories'
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(100), nullable=False)
    
    books = db.relationship('Book', backref='category_ref')

class Book(db.Model):
    __tablename__ = 'books'
    isbn = db.Column(db.String(13), primary_key=True)
    title = db.Column(db.String(255), nullable=False)
    edition = db.Column(db.String(50), nullable=False)
    price = db.Column(db.Numeric(10, 2), nullable=False)
    status = db.Column(db.String(20), default='Available')
    language = db.Column(db.String(50), nullable=False)
    
    publisher_id = db.Column(db.Integer, db.ForeignKey('publishers.id', ondelete='CASCADE'), nullable=False)
    category_id = db.Column(db.Integer, db.ForeignKey('categories.id', ondelete='SET NULL'), nullable=True)
    
    authors = db.relationship('Author', secondary=book_authors, backref=db.backref('books', lazy='dynamic'))

class Member(db.Model):
    __tablename__ = 'members'
    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, db.ForeignKey('users.id', ondelete='CASCADE'), nullable=False, unique=True)
    membership_type = db.Column(db.String(50), nullable=False)
    reg_date = db.Column(db.Date, default=datetime.utcnow)
    expiry_date = db.Column(db.Date, nullable=False)
    
    transactions = db.relationship('Transaction', backref='member_ref', cascade='all, delete-orphan')

class Transaction(db.Model):
    __tablename__ = 'transactions'
    id = db.Column(db.Integer, primary_key=True)
    book_id = db.Column(db.String(13), db.ForeignKey('books.isbn', ondelete='CASCADE'), nullable=False)
    member_id = db.Column(db.Integer, db.ForeignKey('members.id', ondelete='CASCADE'), nullable=False)
    issue_date = db.Column(db.Date, default=datetime.utcnow)
    due_date = db.Column(db.Date, nullable=False)
    return_date = db.Column(db.Date, nullable=True)
    fine = db.Column(db.Numeric(10, 2), default=0.0)
    
    book_ref = db.relationship('Book', backref='transactions')

class PurchaseOrder(db.Model):
    __tablename__ = 'purchase_orders'
    id = db.Column(db.Integer, primary_key=True)
    publisher_id = db.Column(db.Integer, db.ForeignKey('publishers.id', ondelete='CASCADE'), nullable=False)
    order_date = db.Column(db.Date, default=datetime.utcnow)
    total_amount = db.Column(db.Numeric(10, 2), nullable=False)
    status = db.Column(db.String(50), default='Pending')

File: backend/models/Library.js
const mongoose = require('mongoose');

// Publisher Schema
const PublisherSchema = new mongoose.Schema({
  name: { type: String, required: true },
  address: { type: String, required: true },
  contact_email: { type: String, required: true },
}, {
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Author Schema
const AuthorSchema = new mongoose.Schema({
  name: { type: String, required: true },
  biography: { type: String, default: '' },
  date_of_birth: { type: Date, default: null },
}, {
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Category Schema
const CategorySchema = new mongoose.Schema({
  name: { type: String, required: true }
}, {
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Book Schema
const BookSchema = new mongoose.Schema({
  _id: { type: String, required: true }, // We use isbn as _id
  title: { type: String, required: true },
  edition: { type: String, required: true },
  price: { type: Number, required: true },
  status: { type: String, default: 'Available' },
  language: { type: String, required: true },
  publisher_id: { type: mongoose.Schema.Types.ObjectId, ref: 'Publisher', required: true },
  category_id: { type: mongoose.Schema.Types.ObjectId, ref: 'Category', default: null },
  authors: [{ type: mongoose.Schema.Types.ObjectId, ref: 'Author' }]
}, {
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Virtual for ISBN to align with original database schemas
BookSchema.virtual('isbn').get(function() {
  return this._id;
});

// Member Schema
const MemberSchema = new mongoose.Schema({
  user_id: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, unique: true },
  membership_type: { type: String, required: true },
  reg_date: { type: Date, default: Date.now },
  expiry_date: { type: Date, required: true }
}, {
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Transaction Schema
const TransactionSchema = new mongoose.Schema({
  book_id: { type: String, ref: 'Book', required: true }, // refers to book _id (isbn)
  member_id: { type: mongoose.Schema.Types.ObjectId, ref: 'Member', required: true },
  issue_date: { type: Date, default: Date.now },
  due_date: { type: Date, required: true },
  return_date: { type: Date, default: null },
  fine: { type: Number, default: 0.0 }
}, {
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// PurchaseOrder Schema
const PurchaseOrderSchema = new mongoose.Schema({
  publisher_id: { type: mongoose.Schema.Types.ObjectId, ref: 'Publisher', required: true },
  order_date: { type: Date, default: Date.now },
  total_amount: { type: Number, required: true },
  status: { type: String, default: 'Pending' }
}, {
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

const Publisher = mongoose.model('Publisher', PublisherSchema);
const Author = mongoose.model('Author', AuthorSchema);
const Category = mongoose.model('Category', CategorySchema);
const Book = mongoose.model('Book', BookSchema);
const Member = mongoose.model('Member', MemberSchema);
const Transaction = mongoose.model('Transaction', TransactionSchema);
const PurchaseOrder = mongoose.model('PurchaseOrder', PurchaseOrderSchema);

module.exports = {
  Publisher,
  Author,
  Category,
  Book,
  Member,
  Transaction,
  PurchaseOrder
};
